### PR TITLE
Add stack overflow check, Tracealyzer support, and watchdog timer to DIM

### PR DESCRIPTION
**Pull request by thekenu**
### Summary

Add stack overflow check, Tracealyzer support, and watchdog timer to DIM. Along the way, also cleaned up some function names (e.g. Rename functions in `Io_StackOverflow` as `Io` instead of `App`). Also updated Tracealyzer to 4.3.8.

### Testing Done

- Manually tested the correct behaviour of stack overflow check and watchdog timeout
- Tested that Tracealyzer still works on DIM.elf and PDM.elf using DIM hardware

### Checklist
*Please change `[ ]` to `[x]` when you are ready.*
- [ ] I have read and followed the code conventions detailed in [README.md](../README.md) (*This will save time for both you and the reviewer!*).
- [ ] If this pull request is longer then **500** lines, I have provided *explicit* justification in the summary above explaining why I *cannot* break this up into multiple pull requests (*Small PR's are faster and less painful for everyone involved!*).

### CODE REVIEW

File: boards/shared/third_party/TraceRecorder/streamports/Jlink_RTT/include/trcStreamingPort.h [hunk 1]
```diff
@@ -71,7 +71,7 @@ extern "C" {
  * But integrated J-Link OB interfaces are slower and needs about 5-10 KB, 
  * depending on the amount of data produced.
  ******************************************************************************/
-#define TRC_CFG_RTT_BUFFER_SIZE_UP 2000
+#define TRC_CFG_RTT_BUFFER_SIZE_UP 5000
```
Comment (thekenu) [inline boards/shared/third_party/TraceRecorder/streamports/Jlink_RTT/include/trcStreamingPort.h:74]: Now that we have more RAM on the boards we can afford larger buffer for Tracealyzer

**Comment by thekenu** [comment]
Sorry got too eager and did some general cleanups for `SharedStackWaterMark` and `SharedWatchdog`. I've opened a new PR for them at #607. Do you mind reviewing that and then we'll come back to this PR?

**Comment by thekenu** [comment]
@garethellis0 The board-specific changes in this PR is now only for DIM :+1:

File: boards/shared/third_party/TraceRecorder/streamports/Jlink_RTT/include/trcStreamingPort.h (hunk 1 above)
Comment (garethellis0) [reply boards/shared/third_party/TraceRecorder/streamports/Jlink_RTT/include/trcStreamingPort.h:74]: Fine by me, we should just keep this in mind if we ever need to free up RAM, given how much tracelyzer takes up.

**Pull request merged**